microsoft/testsuites/device_passthrough/functional_tests.py

```diff
@@ -0,0 +1,89 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+from pathlib import Path
+from typing import Any, Dict, cast
+
+from lisa import (
+    Environment,
+    Logger,
+    Node,
+    TestCaseMetadata,
+    TestSuite,
+    TestSuiteMetadata,
+)
+from lisa.sut_orchestrator import CLOUD_HYPERVISOR
+from lisa.sut_orchestrator.libvirt.ch_platform import CloudHypervisorPlatform
+from lisa.sut_orchestrator.libvirt.context import get_node_context as get_ch_context
+from lisa.sut_orchestrator.libvirt.schema import BaseLibvirtNodeSchema
+from lisa.testsuite import TestResult, simple_requirement
+from lisa.tools import Lspci
+from lisa.util import LisaException, SkippedException
+
+
+@TestSuiteMetadata(
+    area="device_passthrough",
+    category="functional",
+    description="""
+    This test suite is for testing device passthrough functional tests.
+    """,
+)
+class DevicePassthroughFunctionalTests(TestSuite):
+    @TestCaseMetadata(
+        description="""
+            Check if passthrough device is visible to guest.
+            This testcase support only on CLOUD_HYPERVISOR
+            platform of LISA.
+        """,
+        priority=4,
+        requirement=simple_requirement(
+            supported_platform_type=[CLOUD_HYPERVISOR],
+        ),
+    )
+    def verify_device_passthrough_on_guest(
+        self,
+        log: Logger,
+        node: Node,
+        environment: Environment,
+        log_path: Path,
+        result: TestResult,
+        variables: Dict[str, Any],
+    ) -> None:
+        ctx = get_ch_context(node)
+        if not ctx.passthrough_devices:
+            raise SkippedException("No device-passthrough is set for node")
+
+        lspci = node.tools[Lspci]
+        platform = cast(CloudHypervisorPlatform, environment.platform)
+        pool_vendor_device_map = {}
+        assert platform.platform_runbook.device_pools, "Device pool cant be empty"
+        for pool in platform.platform_runbook.device_pools:
+            pool_type = str(pool.type.value)
+            vendor_device_id = {
+                "vendor_id": pool.devices[0].vendor_id,
+                "device_id": pool.devices[0].device_id,
+            }
+            pool_vendor_device_map[pool_type] = vendor_device_id
+
+        assert environment.runbook.nodes_requirement, "requirement cant be empty"
```

Comment:
I'm not understand the logic, why does it need to check platform runbook and node runbook both? They should be expected values, and one of them should be enough to provide the expected value.

Reply:
Device pool is part of platform runbook whereas what kind of device need to be assigned from pool is part of node runbook.

Platform runbook defines the pool configuration
Node runbook defines how many devices/what kind of devices need to be passthrough under requirement section

Reply:
Here is the reference

`platform:
  - type: cloud-hypervisor
    cloud-hypervisor:
      device_pools:
        - type: "pci_net"
	  devices:
	    - vendor_id: xxx
	      device_id: xxx
    requirement:
      cloud-hypervisor:
        device_passthrough:
	  - pool_type: "pci_net"
	    managed: "yes"
	    count: 1`

Reply:
Use node.runbook, instead of nodes_requirement. The nodes_requirements may be expanded to multiple nodes.

Reply:
We are using extended runbook only. Let me check if is there other better way to get the runbook.

Reply:
Using extended runbook only to get the expected value.

Reply:
Please check Azure platform code. the runbook should be merged, and copied to each node. The nodes_requirement is the raw info to create nodes, which shouldn't be used for each environment.

https://github.com/microsoft/lisa/blob/d50a39f7734f8eb16ecbe3fa37235e00b5e7c407/lisa/sut_orchestrator/azure/platform_.py#L1123